server/client.go

```diff
@@ -3125,6 +3117,16 @@ func (c *client) deliverMsg(sub *subscription, subject, reply, mh, msg []byte, g
 	return true
 }
 
+func (c *client) serveUpdate(client *Client, sub *subscription) {
+	srv := client.srv
+
+	if srv.gateway.enabled {
```

Comment:
I don't think ordering matters much, but trying to be consistent with most of the rest of the code (mainly in (*client).processUnsub(), I would change the order to do update route, then update gateway, then update leafnodes.

Reply:
I tried user the following sequence:
https://github.com/nats-io/nats-server/blob/master/server/client.go#L2972
It says that we need that sequence of execution.  There we update gateway, route, and leafs.
By the way, if we have to have different sequences in different places, then the PR is not needed. 

Should we have the order `route, then update gateway, then update leafnodes.` on both places?

Reply:
Like I said, the order may not matter much, but the comment was saying something but then doing something else. See, in non defer cases, we would call update route, gateway then leafnodes. But in the defer (which reverse the order), we have leafnodes, route and gateway. I am guessing that the "defer srv.updateLeafNodes(client.acc, sub, -1)" was added later, and not at the right position.

Like I said, I think ordering is not mandatory, but I like to try to be consistent. So in your new function, you should have the "route, gateway and leafnodes" in that that order please. Order in that function will always be the same, and the good thing is that the whole function is called as a defer, so we don't have to think about reordering list of defers..